### ISSUE TITLE
Service management: Deployment

### ISSUE THREAD

**Issue by dmolesUC** [CDLUC3/mrt-doc#13]
We want to be able to add/remove/redeploy any instance of any service at
will. Obstacles to that include:

- complexity of deployment/configuration
  - configuration of individual servers (e.g. repository/node directories on storage)
  - configuration of communication between services (mostly load-balanced, sometimes server-locked)
    - what's server-locked (in configuration) besides UNM? UCSB?
    - note that we can probably remove UNM storage early next year
  - per-server Capistrano environment configurations
- complexity of managing load balancers (ALB, ELB, NLB, Apache)
- complexity of startup process
  - order dependence of services (everything depends on storage)
- fragility of long-running processes
- difficulty of determining current state of the system
- lack of orderly shutdown process (no "closing" state that can finish
  existing work without accepting new work)

Note that these same issues also make it difficult or impossible for us to
move to automatic patching / restarts.

## Questions/Action Items

Some long-term goals:

- [blue/green deployment](https://martinfowler.com/bliki/BlueGreenDeployment.html)
- [continuous delivery](https://martinfowler.com/bliki/ContinuousDelivery.html)

Short- to medium-term action items (as much as possible before move to Linux 2):

1. for each service, figure out what's involved in setting up a new instance
   & make a checklist (to some extent these may already exist)
   - system configuration, packages, etc.
   - additional software installation (Tomcat?)
   - Merritt software
     - build
     - deployment
     - configuration
   - Apache and/or ALB
2. figure out *how* we want to add/remove/redeploy services (ideally)
   - some combination of Puppet and Capistrano?
   - target audience: Jim? Mark? Perry?
3. identify places where we assume a certain fixed set of servers
   - load balancers
   - Capistrano environments
   - UNM storage nodes (note: we can probably remove these early next year)
   - ???
4. for each service and server, make sure all (production) configuration is
   under source control
5. for each service:
   - identify obstacles to fully automated setup
   - make a plan to remove those obstacles

(One idea, per Jim: a small 'bootstrap' instance with one of every service)

Longer term action items:

- introduce "closing" state for orderly shutdown
  - may require coordination between services & load balancers?
- when services are unavailable, time out & retry gracefully
- simplify state determination
- figure out how not to lock ingest jobs to servers (more on that below)

Q: Do we need a 'starting' state as well as a 'closing' state?

(Note: we should figure out which of these are either lowest-hanging fruit or most
bang for the buck, and prioritize those)

**Comment by dloy** [comment]
*** Comment Merritt configuration - very manual
Historically configuration for storage, inv, replic and audit have been handled as a combination of a properties file and a compile source (usually private content).

Most of this configuration is out dated - storage still contains d-flat content.

Using file based configuration has the advantage of being easily changed (manually) while relying on the security of the server. It has the huge disadvantage that multiple copies of a service require multiple configurations that are usually redundant.

Moving in the direction of being able to automatically spawn copies of servers requires some mechanism that can easily pass in the needed configurations. 

* Storage configuration (store-info.txt):
- Directory path of each node used by storage

* Node configuration (can-info.txt)
- Node number
- Where is the service
- Bucket information for access
- Whether the content is GET able (Glacier no)

* Inv (inv-info.txt)
- MySQL database information
- Zookeeper information

* Replication:
- Storage service if cloud handling fails
- Inv db information

* Audit
- Storage service if cloud handling fails
- Translation of inv db file URL based on current storage location
-

**Comment by elopatin-uc3** [comment]
Over the past year we've made lots of progress in the area. We've implemented use of a centralized parameter store in almost all microservices (OAI and SWORD left to go). This plays into use of entirely automating the process of spinning up new hosts with Puppet, as is now possible with UI hosts. Additional Puppet automation for other microservices is on the way.

Closing this as much associated work is tracked elsewhere.
